### ISSUE TITLE
[Infrastructure] Add header files to native Visual Studio coreclr projects

### ISSUE THREAD

**Issue by 4creators** [dotnet/coreclr#14884]
Currently CMake is used according to standard pattern where only `.cpp` file are included into targets with library or executable. However, this pattern breaks some usability of CMake created Visual Studio projects since they do not include header files by default and it is necessary to search for them in VS project dependencies directory containing several hundred files or scroll to top of code file and click "open file" on include directive. Obviously none of this is convenient even for medium sized code files.

If proposal will be accepted I could provide PR with required changes.

**Comment by 4creators** [comment]
cc @weshaggard @janvorli

**Comment by RussKeldorph** [comment]
@dotnet/coreclr-contrib @dotnet/jit-contrib @janvorli I may not fully understand, but I assume this would increase the maintenance burden of cmake project files.  Can you guys vote (👍, 👎) on the initial comment for whether you would like @4creators to proceed?

**Comment by janvorli** [comment]
@RussKeldorph you are right it would increase the maintenance burden, but I don't think it would be a significant increase due to the fact that it is very rare to add or remove a header to coreclr. But the issue I have with it is that it would bloat the CMakeLists.txt files. Especially in cases where the list of files compiled depends on target architecture and also whether we build debugger DAC stuff or the target binary. Then the includes would also need to be conditionalized accordingly. See e.g. src/vm/CMakeLists.txt.
So I am torn here, since I can also see the benefit of having the headers listed in the visual studio per project.

**Comment by AaronRobinsonMSFT** [comment]
I am inclined to accept this, but I would hope the generated MSBuild project files use the appropriate `Item` tags. @4creators can you confirm the tags the generated project files uses for these header files?

**Comment by 4creators** [comment]
@AaronRobinsonMSFT 

CMake by default classifies C++ files based on their extension and it means that all header files (`.h`, `.hpp`, `.inl`) are `Header Files` and are not passed to the compiler. Depending on which file types are passed to CMake C++ project additional filters for those file types are added to vcxproj. Existing examples are below:

### clrgc and clrjit projects with headers
![clrgc-clrjit-vcxproj](https://user-images.githubusercontent.com/801959/45916009-09c6e380-be5f-11e8-92ca-3de7916932f3.png)

### clrgc expanded headers
![clrgc-headers-vcxproj](https://user-images.githubusercontent.com/801959/45916013-0f242e00-be5f-11e8-9d7a-0cad4fa313f7.png)

### clrgc expanded sources
![clrgc-sources-vcxproj](https://user-images.githubusercontent.com/801959/45916015-13e8e200-be5f-11e8-97e2-3b473c54608a.png)

**Comment by AaronRobinsonMSFT** [comment]
@4creators What is shown in the Solution explorer isn't actually relevant in this case - at least from a build perspective. That is handled by the `.filters` file and is orthogonal to what is defined in the generated `.vcxproj` file. Can you open the `.vcxproj` and ensure all header files are included using the `ClInclude` tag (e.g. `<ClInclude Include="pch.h" />`)? Thanks.

**Comment by AaronRobinsonMSFT** [comment]
Can we close this now?

**Comment by 4creators** [comment]
@AaronRobinsonMSFT Not yet. There are several projects which are waiting for conversion. I am doing it step by step.

**Comment by janvorli** [comment]
@4creators I think you have converted all of the projects already and the issue can be closed, right?

**Comment by 4creators** [comment]
@janvorli There are still 3 small ones to be converted. I will try to go back to them in September if that's OK.

**Comment by janvorli** [comment]
@4creators of course! Thank you.
